### DISCUSSION TITLE
RFC: next/form - Interactions for Next.js

### DISCUSSION THREAD

**Discussion by lipp** [vercel/next.js#13077] [Ideas]
# Motivation

Next.js has all the batteries included for dealing with the "passive" side of a web application. However, when it comes to user interaction beyond "clicking links" Next.js feels pretty naked.

IMHO Next.js can be completed with small and optional pieces to allow truly interactive web applications, including server-side data manipulation.

The proposed technique employs the very well proven [post-redirect-get pattern](https://en.wikipedia.org/wiki/Post/Redirect/Get), while benefitting from the speed of Next.js.

Just like `<next/link>`, the implementation is `<noscript>` compatible, proving the unopinionated approach.

# Summary

```tsx
import Form from 'next/form'
```

A `Form` component which behaves just like the native one, but benefits from client-side navigation and fast view refresh, like `next/link` does.

Check out [kimeta.de](https://www.kimeta.de) which uses this approach extensively.

The component enables **search forms** via simple navigation and **mutations** via [post-redirect-get](https://en.wikipedia.org/wiki/Post/Redirect/Get) pattern without any additional JavaScript code. It enables full interactivity with declarative components. Just like `next/link` it blends in seamlessly with native behaviour and even works with JS disabled. 

Supports:
 - controlled inputs
 - uncontrolled inputs
 - all input types

# Goals

Enable _pure declarative_ user input. No additional (client) JS code needed.

This can actually make additional app state management obsolete and makes it easy to stay consistent between reloads etc. 

# 1. Search forms

A very basic search form could look like this:

```jsx
<Form>
  <input name="q" type="search" />
  <input type="submit" />
</Form>
```
Just like a native `form` would navigate to current pathname with query `q` on submit event, `Form` does the very same automatically,  employing `next/router`. The form's `method` defaults to "get". 

In this case the `<input>`s are uncontrolled. But controlled inputs with custom validation or auto-complete are also supported.

# 2a. Mutations (client) 

A very basic form to perform a mutation could look like this.

```jsx
<Form method="post" action="/mutation/set-age" redirectUrl={{pathname: '/foo', query: {bar: 1}}} >
   <input name="age" type="number" />
   <input name="user-id" type="hidden" value="123" />
   <input type="submit" />
</Form>
```

On submit, the form data will be automatically extracted and posted to **action** "/mutations/set-age" and a redirect to **redirectUrl** "/foo?bar=1&**mr-name=set-age&mr-success=1**" in case of success will take place. The additional query params are:
  - `mr-name` the name of the mutation 
  - `mr-success=1` success or no success

`mr-name` and `mr-success` allow the display of feedback like messages / snackbars etc. . _Maybe this should be JSON.stringified response of the mutation?_

# 2b. Mutations (server)

Every mutation should be placed in particular directory, e.g. `/pages/mutations`. By placing the file in this directory, a special `mutation` wrapper is applied which deals with the redirect logic and augments the query with `mr-name` and 
`mr-success` on redirect. Further, it extracts the form parameters. It can deal with no-JS (native) form post data and the mimicked client-side request. 

```ts
import {FormParameters} from 'next/form'
export default async (req: NextApiRequest, res: NextApiResponse, formParams: FormParameters) => {
  // perform the mutation
  // eventually throw exception
}
```

# `<Form>` props

```tsx
type Props = React.ComponentProps<'form'> & {
	hiddenInputs?: ParsedUrlQueryInput // { [key: string]: string | number | undefined | boolean }
	scrollTop?: boolean
	redirectUrl?: UrlObject // required if method === 'post'
    }
```

# How does it work?

The `Form` component renders a native `form` and forwards all the props. Additionally, it attaches `onSubmit` which does two different things, depending on `method`.

## Search form / method = 'get' (default)

`action` in this case is a "normal" page url.

 - collects the form data
 - `Router.push(/*action + form data as query params */)`
 - page "refresh" (just ordinary Next.js navigation)
 - optionally scroll to top

## Mutation / method = 'post'

`action` in this case is a *mutation*.

- collects the form data
- `Router.replace(/*action + form data as query params */)`
-  mutation gets executed
- redirect to `redirectUrl` 

# Implementation Status

I created a fully working version of the `Form` component. It can be implemented very lightweight < 150 lines of code. For the mutation handler I created a "wrapper" `mutation`, which uses the `Page.getInitialProps` API to deal with the redirecting from client and server. This is not ideal, since mutations would ideally be "server-only" code, just like `pages/api` or `getServerSideProps` can do. 

Check out [kimeta.de](https://www.kimeta.de) which uses this approach for most interactions. As this approach still allows additional state, you can see things like auto-complete or loading indicators. You can actually disable JS and try to interact (search, bookmark etc).

**Comment by lipp** [comment]
### Alternative 1

This could be an alternative component API:

```jsx
<Interaction action="..." method="..." redirectUrl="..." >
  <form>
    ...
  </form>
</Interaction>
```

This would have the advantage of working better with `styled-jsx` if you want to style the form or use it in selectors.

**Comment by lipp** [comment]
### Alternative 2

Or two different components for search and mutation

```jsx
import {Search, Mutation} from 'next/form'

// form.method will be set to "get"
<Search>
  <form>
    ...
  </form>
</Search>

// form.method will be set to "post"
<Mutation action="..." redirectUrl="..." >
  <form>
    ...
  </form>
</Mutation>

**Comment by lipp** [comment]
@timneutkens @ijjk @Timer any opinion on this? Feedback would be very welcomed 🙏

**Comment by lipp** [comment]
# Usage

**I created a [Sandbox]((https://codesandbox.io/s/next-form-q8u18?file=/pages/index.tsx))**, with the current implementation and two examples:

1. [A search form](https://codesandbox.io/s/next-form-q8u18?file=/pages/index.tsx)

Features a stateless search form with uncontrolled input. The most important [lines of code for the client](https://codesandbox.io/s/next-form-q8u18?file=/pages/index.tsx).

```tsx
   // in pages/index.tsx
   <Interaction>
      <form>
        <input
          type="text"
          name="q"
          placeholder="Enter search term"
          defaultValue={q || undefined}
        />
        <input type="submit" />
      </form>
    </Interaction>
```

The `<Interaction>` components attaches an `onSubmit` handler to the form, which behaves as a _noscript_ form, except it uses `next/router`.

2. [A Mutation Form](https://codesandbox.io/s/next-form-q8u18?file=/pages/profile.tsx)

Features a stateless form with uncontrolled input for data manipulation on "the server". The age form data is validated on the server. A super simple `<Snackbar >` component is employed to demonstrate the usage of the mutation`s response.

The most important [lines of code for the client](https://codesandbox.io/s/next-form-q8u18?file=/pages/profile.tsx):

```tsx
   // in pages/profile.tsx
   <Interaction>
      <form method="post" action="/mutations/set-profile">
        <input
          type="text"
          name="name"
          placeholder="Your name"
          defaultValue={profile.name}
          required
        />
        <input
          type="number"
          name="age"
          placeholder="Your age"
          defaultValue={profile.age}
        />
        <input type="submit" />
      </form>
    </Interaction>
```

The most important [lines of code for the server](https://codesandbox.io/s/next-form-q8u18?file=/pages/mutations/set-profile.ts):

```ts
 // in pages/mutation/set-profile.ts
export default mutation(async (ctx, formParams) => {
  try {
    await setProfile(ctx.req, {
      name: formParams.name as string,
      age: parseInt(formParams.age as string, 10) || undefined
    });
    return {
      result: "ok"
    };
  } catch (err) {
    return {
      result: err.toString()
    };
  }
});
```

The `<Interaction>` components attaches an `onSubmit` handler to the form, which behaves as a _noscript_ form, except it uses `next/router` with a POST-REDIRECT-GET pattern.

# Implementation

## Client

The `<Interaction>` component is defined in [`next-form/Interaction`](https://codesandbox.io/s/next-form-q8u18?file=/next-form/Interaction.tsx). Not sure if the `autoFocus` stuff makes sense. Maybe there is a way to have less code. 

The component `<HiddenInputs>` is not used in the examples yet, but is intended as a helper if you have to provide hidden inputs as context for your mutation coming from a `ctx.query / Router.query`.

```tsx
<Interaction>
  <form method="post" action="/mutations/set-bookmark">
    <button type="submit">Toggle</button>
    <HiddenInputs hiddenInputs={{id: itemId, ...query}} />
  <form>
</Interaction>
```

## Server

The implementation for the server is in [`next-form/mutation.ts`](https://codesandbox.io/s/next-form-q8u18?file=/next-form/mutation.ts). It is "just wrapping up" a `NextPage`. It collects the form data and calls the wrapped mutation function. It then redirects to `redirectUrl` or to the current page and adds the response of the mutation to the query to enable feedback like messages/snackbars.

@timneutkens @Timer @ijjk and community: Any feedback is very appreciated :)

**Comment by leerob** [comment]
Looks interesting. I'm struggling to understand what it's used for. Could you explain what real-world problem this attempts to solve with some examples? 🙏 

This would help explain why it should be included in the framework.

**Comment by lipp** [comment]
@leerob thanks for your interest! I'll give my best to elaborate on this idea:

The proposal tries to provide a unified, simpler and less error prone way of dealing with user interactions. In my opinion, there are three kinds of user interactions:
 1. changes of a "view" triggered by "single click" (typical link) => `<a>` => `next/link`
 2. changes of a "view" triggered by "user input" (search form) => `<form method="get">` => ???
 3. mutations of data triggered by "user input"  (form) => `<form method="post">` => ???

Whereas 1. is perfectly addressed by `<next/link>` fully embracing native browser behaviour, 2. and 3. are not addressed by Next.js at all. There are endless possibilities to deal with that, in many cases involving collecting input state in the client and then performing "fetch" / "ajax". One tricky part is, to keep client and "server" state in sync. Some implementations rely on being optimisitc, while others employ manual imperative routing or relying on client side state management altogether (this is the error prone part). 

The proposal focuses on the basics and emulates, how No-JS 1999 like frameworks did work. The proposal also encourages the use of standard forms and named inputs which have worked for decades. By using `<Interaction>`, no additional client JS code is required for many scenarios.

At the same time, it imposes no restrictions on using JS as progressive enhancement (like autocomplete etc).

So coming back to real-world problem: I guess almost any web application has user interactions of type 2. and 3. But up to now, there is no go-to / default way to solve this.

**Comment by lipp** [comment]
Actual examples of usage of this approach can be found on kimeta.de

1. Search form on index and search pages
2. Bookmark / Heart buttons (actually are submit buttons with hidden Inputs)
3. Editing folders of the bookmarked items 

You can disable JS on the page and interactions still work.

@leerob did my answer clear things up? Please don’t hesitate to ask anything or to dig deeper.

**Comment by lipp** [comment]
@Timer @timneutkens and community: does this idea make any sense to you? Your short feedback would be really appreciated !

**Reply by lipp** [reply]
Ping @Timer @timneutkens @leerob  . Any chance you can comment on this?

**Comment by revskill10** [comment]
I don't understand why some of authors prefer the old school way of `form` where we don't need AJAX.

In most real world use cases, AJAX is a must-have, especially in an enterprise application.
For example, suppose we're in the `create new order` form, where we could add new `order line item`.
Now we want to change the `pricelist` for all of those new order line items (they're not saved yet), without AJAX, how do you suppose to resolve these kinds of use cases ?

So, my suggestion is, make AJAX first-class support in a form component where interaction between client and server is complicated.

**Reply by jnv** [reply]
Similar to next/link, nothing prevents you from implementing more complex, custom, AJAX-driven forms. But from my experience working with junior and mid-level front-end developers, there is a tendency to overuse AJAX even for simple forms. Some devs end-up reimplementing most of the native forms behavior in poor way, usually due to lack of experience and/or lack of established best practices.

Also many "traditional" server-side web frameworks like Ruby on Rails have robust solutions for handling form submissions, with built-in validation and CSRF protection. Yet in Node.js land it seems to me this is still an absolute PITA where the client-side and server-side logic for form submission is usually completely decoupled, causing duplication and subtle bugs.

So I for one welcome an attempt to provide a sensitive and unobtrusive forms in Next with first-class support for server-side logic. However I do think this could be a separate community plugin to better explore various use cases and test the API "in the wild".

**Reply by revskill10** [reply]
One of the reason i use REACT is because of ease of componentation and doing ajax on a complex form.
I used Rails before, for the view and server-side form, but i have to remove it and use React for frontend instead.
So to me, the reason i still use React is to provide user the best experience as possible when working with complex form, so if one needs "simple form", React is not needed actually.

**Reply by jnv** [reply]
Well, while I agree that React is _in general_ overkill for simple forms, it's not really a choice for an all-encompassing framework like Next.js where React basically owns your whole website.

So the discussion here really isn't whether you should use React for simple forms, that's already given.

**Reply by lipp** [reply]
Providing working form / input primitives is just as basic and fundamental as having links on a website. Not providing it, forces every developer for interactive websites to do their own thing. Eventually pulling in bloaty libs, which negatively impact bundle sizes. 

The `noscript` compatibility proves, that the approaches above respects webstandards. The proposal will improve in-client performance / UX just as `next/link` does. In addition, devs can still use progressive enhancement or completely opt-out to use custom interaction logic. 

As almost any website sooner or later needs some kind of interactivity beyond links and as there is an easy way to implement this at close to 0 bytes for the front end, this should be part of Next.js.

**Comment by KATT** [comment]
I've hacked together a working SSR form implementation that has end-to-end type safety & data validation. Ended up turning it into a library that actually works. Still has some rough edges but it works both with and without JavaScript enabled in the browser. 

Have a look and feel free to setup issues/discussions about how it could be improved. https://github.com/KATT/next-ssr-form

Reference usage: https://github.com/KATT/next-ssr-form/blob/main/examples/typescript/pages/formik-scaffold.tsx or https://github.com/KATT/next-ssr-form/blob/main/pages/index.tsx

**Comment by jneander** [comment]
Has there been any discussion elsewhere about no-js form submission in Next.js? When my coworkers first pitched using Next.js, it sounded like this was behavior provided out of the box. But I must have misunderstood.

**Reply by jnv** [reply]
I am not aware of such discussion, a research on progressively enhanced forms in Next.js actually lead me to this thread.
